### PR TITLE
Added Employee creation, deletion, and editing

### PR DESCRIPTION
**Pull request by ryannoort**

### CODE REVIEW

**Comment by jfeaver** [comment]
Please add `.DS_Store` to your `.gitignore_global` file and then remove those files from the repository. Do you know what the `MrClean` submodule is? My guess is that it is a git submodule but I'm really not sure. At any rate, we don't need it and it should be removed from the repository.

File: app/views/employee/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<h1>Employees</h1>
+
+<table>
+	<tr>
+		<th>Employee Name</th>
+	</tr>
+	<tr>
+	<% emp = Employee.all.order(:first_name) %>
+	<% emp.each do |e| %>
```
Comment (jfeaver) [inline app/views/employee/index.html.erb]: We should use a full variable name `employee`. It's more typing for you at first but it is easier to read and much  faster to comprehend for you and others in the future.

File: app/views/employee/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+<h1> Add New Employee </h1>
+
+<%= form_for :employee do |form| %>
+  <p>
```
Comment (jfeaver) [inline app/views/employee/new.html.erb]: divs over ps

File: app/views/employee/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<h1><%= "#{@employee.first_name} #{@employee.last_name}" %></h1>
```
Comment (jfeaver) [inline app/views/employee/show.html.erb]: If you'd like, you can create a presenter object so that you can do `@employee.full_name`.

Here's a simple solution: http://nithinbekal.com/posts/rails-presenters/
And here's a little bit more full-featured way to go (but more time and more learning): https://github.com/fnando/burgundy

File: app/views/schedule/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<h1>Kitchen Schedule</h1>
+
+<table>
+
+<% schedule = Schedule.all %>
```
Comment (jfeaver) [inline app/views/schedule/index.html.erb]: move this to the controller.

File: app/views/assignment/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+<h1>Kitchen Schedule</h1>
+
+<table>
+
+<% @weeks.each do |w| %>
+  <tr>
+    <td>
+      <%= "#{w.starts_at.strftime('%B %d, %Y')}" %>
+    </td>
+    <% @assignments.where(week: w).each do |a| %>
```
Comment (jfeaver) [inline app/views/assignment/index.html.erb]: Expand the single letter variables.

This line should be `week.employees.each do |employee|`

File: app/models/week.rb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+class Week < ActiveRecord::Base
+  validates_presence_of :starts_at
+  validates_uniqueness_of :starts_at
+
+  has_many :assignments
```
Comment (jfeaver) [inline app/models/week.rb]: Also, `has_many :employees, through: :assignments`.

File: config/routes.rb [hunk 7]
```diff
@@ -1,3 +1,14 @@
 Rails.application.routes.draw do
+
+  resources :employee
+  resources :assignment
+  resources :week
+
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  get  '/schedule', to: 'assignment#index'
+  get  '/employee', to: 'employee#index'
+  post '/employee/new', to: 'employee#create'
```
Comment (jfeaver) [inline config/routes.rb]: These three routes should be covered by the resources routes above.

File: config/routes.rb [hunk 8]
```diff
@@ -1,3 +1,14 @@
 Rails.application.routes.draw do
+
+  resources :employee
+  resources :assignment
+  resources :week
```
Comment (jfeaver) [inline config/routes.rb]: I prefer to use `only: [:index, :new, :create]` or whatever actions if I'm ever not using all of the CRUD actions.

File: db/seeds.rb [hunk 9]
```diff
@@ -5,3 +5,39 @@
...
+if Rails.env.development?
+  #fills employees table with dummy employees
+  Employee.create(first_name: "Aaron", last_name: "Andrews", email: "Aaron@mail.com", phone: "1-800-555-1111")
+  Employee.create(first_name: "Bret", last_name: "Bean", email: "Bret@mail.com", phone: "1-800-555-2222")
+  Employee.create(first_name: "Charlie",last_name: "Chase", email: "Charlie@mail.com", phone: "1-800-555-3333")
+  Employee.create(first_name: "Dave", last_name: "Daniels", email: "Dave@mail.com", phone: "1-800-555-4444")
+  Employee.create(first_name: "Eric", last_name: "Ericson", email: "Eric@mail.com", phone: "1-800-555-5555")
+
+  #fill week table with 5 week starting Sept 4, 2017
+  @dt = DateTime.new(2017,9,4).beginning_of_week
```
Comment (jfeaver) [inline db/seeds.rb]: I'd probably call the `@current_time`.

File: db/seeds.rb [hunk 10]
```diff
@@ -5,3 +5,39 @@
...
+  
+  for i in 0..10
```
Comment (jfeaver) [inline db/seeds.rb]: also: `10.times do |i|`

File: app/controllers/assignment_controller.rb [hunk 11]
```diff
@@ -0,0 +1,9 @@
+class AssignmentController < ApplicationController
+  protect_from_forgery with: :exception
```
Comment (jfeaver) [inline app/controllers/assignment_controller.rb]: I'd put this in `ApplicationController` or in a new controller called `AdminController` since it will only be relevant for pages that an admin handles (all the forms for creating and updating).

File: app/controllers/employee_controller.rb [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    @employee.update_attributes(employee_params(params))
+
+    redirect_to @employee
+  end
+
+  def destroy
+    @employee.destroy
+    flash[:success] = "Employee Deleted"
+
+    redirect_to employee_index_path
```
Comment (jfeaver) [inline app/controllers/employee_controller.rb]: Since you're in the `EmployeeController`, you have the option of doing `redirect_to(:index)`. I like that option here because it means I don't have to think about where the path helper is going. I know I just have to scroll up to see where it goes.

File: app/models/week.rb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+class Week < ActiveRecord::Base
+  validates_presence_of :starts_at
+  validates_uniqueness_of :starts_at
```
Comment (jfeaver) [inline app/models/week.rb]: I'd probably do `validates :starts_at, presence: true, uniqueness: true`.

File: app/views/assignment/index.html.erb [hunk 14]
```diff
@@ -0,0 +1,20 @@
+<h1>Kitchen Schedule</h1>
+
+<table>
+
+<% @weeks.each do |w| %>
+  <tr>
+    <td>
+      <%= "#{w.starts_at.strftime('%B %d, %Y')}" %>
```
Comment (jfeaver) [inline app/views/assignment/index.html.erb]: I might use a presenter method here that overrides `starts_at` to use `week.starts_at` in the view for the same behaviour. In the helper I'd use the `l` method (short for `localize`) that translates timestamps into a locale-specific format. In this case, you could improve this without a presenter by using the `l` helper directly here or by creating a helper method (maybe inside `app/helpers/timestamp_helpers.rb`) so you could do `readable_date(week.starts_at)`. Finally, I'd use `'%B %e, %Y'` because I think leading zeroes look funny in most contexts.

This is an optional change.

File: app/controllers/week_controller.rb [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+  include AssignmentGenerationHelper
+
+  def new
+    last_week = Week.order(:starts_at).last #the last week added to the db
+
+    new_week = last_week.starts_at.next_week.beginning_of_week
+
+    week = Week.create(starts_at: new_week)
+
+    generate_new_week_assignments(week)
```
Comment (jfeaver) [inline app/controllers/week_controller.rb]: I like the approach here to separate complicated business logic from the controller. A better tool for this than a helper method is an object. Please create an object that duplicates the behaviour you're getting with the helper but uses this line of code: `GenerateNewWeekAssignments.new(week: week).call`.

An object is nicer because we can create as many methods as we need to accomplish the task without adding those methods to the controller. Helpers are also added to all views and we don't want views to be generating database records.

File: config/routes.rb [hunk 16]
```diff
@@ -1,3 +1,14 @@
...
+
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  #get  '/schedule', to: 'assignment#index'
+  #get  '/employee', to: 'employee#index'
+  post '/employee/new', to: 'employee#create'
```
Comment (jfeaver) [inline config/routes.rb]: We should remove this. Let me know if it's giving you troubles. The conventional route here (covered by line 3 above) is `POST /employees`.

File: config/routes.rb [hunk 17]
```diff
@@ -1,3 +1,14 @@
...
+  get '/', to: 'assignment#index'
```
Comment (jfeaver) [inline config/routes.rb]: Rails has a special route helper for this if you want to use it: `root to: 'assignment#index'`.

File: db/seeds.rb [hunk 18]
```diff
@@ -5,3 +5,47 @@
...
+  
+  20.times do |i|
+    Week.create(starts_at: @current_time)
+    @current_time = @current_time.next_week.beginning_of_week 
+  end
+
+  #assigns two employees to a week in numerical order
+  @emp_id = 1
+
+  Week.all.each do |w|
```
Comment (jfeaver) [inline db/seeds.rb]: This is fine here but I should point out the Rails' `#find_each` method. If you have 1000s of records in a database then `.all.each` will crash your server when it loads them all into memory. `find_each` loops through batches of records at a time.

**Comment by ryannoort** [comment]
Trello Stories completed. 

- As the manager, I want to enter employee information, so that they can be easily added to the schedule.

- As a manager, I want to be able to create a schedule so that the kitchen duty can be assigned to the employees.

- As an manager, I want to edit the schedule, so that if an employee can't do it when they're scheduled, I can switch them with someone else.

- As an employee, I want to view the schedule, so that I know when I am on kitchen duty.

File: app/classes/generate_new_week_assignments.rb [hunk 19]
```diff
@@ -1,5 +1,5 @@
-module AssignmentGenerationHelper
-  def generate_new_week_assignments(week)
+class GenerateNewWeekAssignments
```
Comment (jfeaver) [inline app/classes/generate_new_week_assignments.rb]: Excellent!

Let's move this from `app/classes/` which is too vague to be a helpful namespace to `app/services/` (some devs might use `app/interactors/`).

File: app/classes/generate_new_week_assignments.rb [hunk 20]
```diff
@@ -1,5 +1,5 @@
...
+  def new(week)
```
Comment (jfeaver) [inline app/classes/generate_new_week_assignments.rb]: The `new` method is like a keyword method and should never be defined in a Ruby class. We use `initialize` instead which is called by `new`. The `initialize` method should only set up the object and we should have a `call` method that does the real work of finding employees and creating assignments.

Finally you should extract some of the code here into private methods in accordance with the single responsibility principle (I think this is in the first or second chapter of that Ruby book 😄). The named private methods will help me understand what your code is doing (it's hard to follow without digging into the logic).

File: app/helpers/generate_new_week_assignments.rb [hunk 21]
```diff
@@ -0,0 +1,27 @@
+class GenerateNewWeekAssignments
```
Comment (jfeaver) [inline app/helpers/generate_new_week_assignments.rb]: We should remove this file but you probably already knew that 😄.

File: config.ru [hunk 22]
```diff
@@ -2,4 +2,6 @@
 
 require_relative 'config/environment'
 
+require "#{RAILS.root}/app/classes/generate_new_week_assignments.rb"
```
Comment (jfeaver) [inline config.ru]: Rails should `autoload` (a Rails' magical method) everything in the `/app` directory so you don't need to require it. If you did need to require it then you'd want to put this kind of line in `config/application.rb` at some point after the `Bundler.require` (It's a fun exercise to follow the boot order in that file).

File: app/classes/generate_new_week_assignments.rb [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+    employee = last_assigned.first.employee
+
+    Assignment.create(week: week, employee: employee)
+
+    last_assigned.shift
+
+    employee = last_assigned.first.employee
+
+    Assignment.create(week: week, employee: employee)
+  end
```
Comment (jfeaver) [inline app/classes/generate_new_week_assignments.rb]: The `new` method is like a keyword method and should never be defined in a Ruby class. We use initialize instead which is called by new. The initialize method should only set up the object and we should have a call method that does the real work of finding employees and creating assignments.

Finally you should extract some of the code here into private methods in accordance with the single responsibility principle (I think this is in the first or second chapter of that Ruby book 😄). The named private methods will help me understand what your code is doing (it's hard to follow without digging into the logic).

File: app/classes/generate_new_week_assignments.rb [hunk 24]
```diff
@@ -0,0 +1,27 @@
+class GenerateNewWeekAssignments
```
Comment (jfeaver) [inline app/classes/generate_new_week_assignments.rb]: Excellent!

Let's move this from app/classes/ which is too vague to be a helpful namespace to app/services/ (some devs might use app/interactors/).

File: app/services/generate_new_week_assignments.rb [hunk 25]
```diff
@@ -1,32 +1,41 @@
...
+          @last_assignment_list.push(last_employee_assignment)
         else
           #new employee who haven't been assigned yet dealt with here
         end
       end
     
-      last_assigned = last_assigned.sort_by {|assignment| [assignment.week, assignment.employee]}
-
-      employee = last_assigned.first.employee
+      @last_assignment_list = @last_assignment_list.sort_by {|assignment| [assignment.week.created_at, assignment.employee.id]}
```
Comment (jfeaver) [inline app/services/generate_new_week_assignments.rb]: Good!

File: app/services/generate_new_week_assignments.rb [hunk 26]
```diff
@@ -1,32 +1,41 @@
...
+    @week = week
+  end
+
+  def call
+    generate_last_assignment_list
+
+    2.times do
+      create_week_assignment
+    end
   end
```
Comment (jfeaver) [inline app/services/generate_new_week_assignments.rb]: Awesome!

File: app/services/generate_new_week_assignments.rb [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+    def generate_last_assignment_list
+      @last_assignment_list = []
+
+      Employee.all.each do |employee|
+        last_employee_assignment = Assignment.where(employee_id: employee.id).order(:week_id).last
+
+        if last_employee_assignment.present?
+          @last_assignment_list.push(last_employee_assignment)
+        else
+          #new employee who haven't been assigned yet dealt with here
```
Comment (jfeaver) [inline app/services/generate_new_week_assignments.rb]: I guess this is a placeholder/TODO item for later? Is it in Trello?

Comment (ryannoort) [reply app/services/generate_new_week_assignments.rb]: I'm not sure how I'm going to do this yet, so I left it blank for now.

**Comment by jfeaver** [comment]
🎉

🌮 🕐
